### PR TITLE
Add model required to  threshold_scan

### PR DESCRIPTION
**Pull request by krautech**
## Description

#164 

image

### CODE REVIEW

File: scanner.py [hunk 1]
```diff
@@ -329,6 +329,13 @@ def __init__(self, config: ConfigWrapper):
             desc=self.cmd_Z_OFFSET_APPLY_PROBE_help,
         )
 
+    def is_model(self, gcmd: GCodeCommand):
```
Comment (Jomik) [inline scanner.py:332]:
```suggestion
    def is_model_loaded(self, gcmd: GCodeCommand):
```

File: scanner.py [hunk 2]
```diff
@@ -329,6 +329,13 @@ def __init__(self, config: ConfigWrapper):
...
+        model = self.model
+        if model is None:
```
Comment (Jomik) [inline scanner.py:334]:
```suggestion
        if self.model is None:
```

File: scanner.py [hunk 3]
```diff
@@ -329,6 +329,13 @@ def __init__(self, config: ConfigWrapper):
...
+            raise gcmd.error(
+                "No model currently selected - make sure you have run CARTOGRAPHER_CALIBRATE first"
```
Comment (Jomik) [inline scanner.py:336]: Should this link to the docs like we do for the validate?

File: scanner.py (hunk 1 above)
Comment (Jomik) [reply scanner.py:332]: We could also validate the model here. Just name the method appropriately 😄

**Review by Jomik — CHANGES_REQUESTED** [review]

File: scanner.py (hunk 3 above)
Comment (krautech) [reply scanner.py]: Done

File: scanner.py (hunk 1 above)
Comment (krautech) [reply scanner.py]: https://github.com/Cartographer3D/cartographer-klipper/pull/165/commits/15803433ec8a043d70c19f21589482b8cb8f6fa7

File: scanner.py [hunk 4]
```diff
@@ -329,6 +335,15 @@ def __init__(self, config: ConfigWrapper):
             desc=self.cmd_Z_OFFSET_APPLY_PROBE_help,
         )
 
+    def is_valid_model_loaded(self, gcmd: GCodeCommand):
+        if self.model is None:
+            raise gcmd.error(
+                "No model currently selected - make sure you have run CARTOGRAPHER_CALIBRATE first."
```
Comment (Jomik) [inline scanner.py:341]:
```suggestion
                f"No model currently selected - make sure you have run {format_macro('CARTOGRAPHER_CALIBRATE')} first."
```

**Review by Jomik — CHANGES_REQUESTED** [review]

File: scanner.py [hunk 5]
```diff
@@ -329,6 +335,15 @@ def __init__(self, config: ConfigWrapper):
...
         )
 
+    def is_valid_model_loaded(self, gcmd: GCodeCommand):
+        if self.model is None:
+            raise gcmd.error(
+                f"No model currently selected - make sure you have run {format_macro('CARTOGRAPHER_CALIBRATE')} first."
+                + f" Click <a href='{DOCS_TOUCH_CALIBRATION}'>HERE</a> for more information"
+            )
+        else:
+            self.model.validate()
```
Comment (Jomik) [inline scanner.py:345]:
```suggestion
        self.model.validate()
```

File: scanner.py [hunk 6]
```diff
@@ -757,11 +770,23 @@ def cmd_SCANNER_THRESHOLD_SCAN(self, gcmd: GCodeCommand):
...
+            or "z" not in kin_status["homed_axes"]
+        ):
             self.trigger_method = TriggerMethod.SCAN
-            raise gcmd.error("Must home X and Y axes first")
+            raise gcmd.error("Must home all axes first")
 
-        # Set initial scan values
+            # Check if model is loaded
+        self.is_valid_model_loaded(gcmd)
+        # Check if bed leveling has been applied
```
Comment (Jomik) [inline scanner.py:783]:
```suggestion
```

File: scanner.py [hunk 7]
```diff
@@ -757,11 +770,23 @@ def cmd_SCANNER_THRESHOLD_SCAN(self, gcmd: GCodeCommand):
...
+            "x" not in kin_status["homed_axes"]
+            or "y" not in kin_status["homed_axes"]
+            or "z" not in kin_status["homed_axes"]
+        ):
             self.trigger_method = TriggerMethod.SCAN
-            raise gcmd.error("Must home X and Y axes first")
+            raise gcmd.error("Must home all axes first")
 
-        # Set initial scan values
+            # Check if model is loaded
```
Comment (Jomik) [inline scanner.py:781]:
```suggestion
```

File: scanner.py [hunk 8]
```diff
@@ -329,6 +335,15 @@ def __init__(self, config: ConfigWrapper):
             desc=self.cmd_Z_OFFSET_APPLY_PROBE_help,
         )
 
+    def is_valid_model_loaded(self, gcmd: GCodeCommand):
```
Comment (Jomik) [inline scanner.py:338]:
```suggestion
    def validate_model_loaded(self, gcmd: GCodeCommand):
```
I'd change the name `is_` makes it sound like a boolean that you have to react to, this is more a statement.
"Is a valid model loaded? Yes - ..., No - ..."
versus
"Validate that a model is loaded" - more of a command, doesn't require a response.

**Review by Jomik — CHANGES_REQUESTED** [review]

File: scanner.py (hunk 8 above)
Comment (krautech) [reply scanner.py:338]: Yeah makes sense, it originally was a bool :D

**Review by Jomik — DISMISSED** [review]

**Review by Jomik — APPROVED** [review]

**Pull request merged**